### PR TITLE
[AMDGPU][ISel] Set trunc store action to expand for v4f32->v4bf16

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp [hunk 1]
```diff
@@ -316,6 +316,7 @@ AMDGPUTargetLowering::AMDGPUTargetLowering(const TargetMachine &TM,
   setTruncStoreAction(MVT::v2f32, MVT::v2f16, Expand);
   setTruncStoreAction(MVT::v3f32, MVT::v3f16, Expand);
   setTruncStoreAction(MVT::v4f32, MVT::v4f16, Expand);
+  setTruncStoreAction(MVT::v4f32, MVT::v4bf16, Expand);
```
Comment (changpeng) [inline llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:319]: Can you also look at other vector size like v2bf16, v8bf16, etc for TruncStoreAction? Thanks.

Comment (shiltian) [reply llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:319]: Yeah, I was about to ask if we want to add them all. Okay will do.

**Review by rampitec — APPROVED** [review]
The produced code is huge IMO, given the simplicity of the operation, but this is not a problem of the vector expand action itself.

**Comment by shiltian** [comment]
> The produced code is huge IMO, given the simplicity of the operation, but this is not a problem of the vector expand action itself.

How can we improve it?

File: llvm/test/CodeGen/AMDGPU/fp_trunc_store_bf16.ll [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+; CHECK-NEXT:    global_store_dwordx4 v[32:33], v[8:11], off offset:32
+; CHECK-NEXT:    global_store_dwordx4 v[32:33], v[0:3], off offset:16
+; CHECK-NEXT:    global_store_dwordx4 v[32:33], v[4:7], off
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %conv = fptrunc <32 x float> %num to <32 x bfloat>
+  store <32 x bfloat> %conv, ptr addrspace(1) %p, align 8
+  ret void
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/fp_trunc_store_bf16.ll:426]: Do the f64 cases work too?

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/fp_trunc_store_bf16.ll:426]: They need to be expanded as well. I'll do it in a follow-up patch.

Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/fp_trunc_store_bf16.ll:426]: https://github.com/llvm/llvm-project/pull/90493

**Pull request merged**